galoisAvx512_amd64.go

```diff
@@ -0,0 +1,181 @@
+//+build !noasm
+//+build !appengine
+//+build !gccgo
+
+// Copyright 2015, Klaus Post, see LICENSE for details.
+// Copyright 2019, Minio, Inc.
+
+package reedsolomon
+
+//go:noescape
+func _galMulAVX512Parallel82(in, out [][]byte, matrix *[(16 + 16) * dimIn * dimOut82]byte, addTo bool)
+
+//go:noescape
+func _galMulAVX512Parallel84(in, out [][]byte, matrix *[(16 + 16) * dimIn * dimOut84]byte, addTo bool)
+
+const dimIn = 8
+const dimOut82 = 2
+const dimOut84 = 4
+
+var zeroCoeff = [16]byte{}
+
+// Construct block of matrix coefficients for 2 outputs rows in parallel
+func setupMatrix82(matrixRows [][]byte, inputOffset, outputOffset int, matrix *[(16 + 16) * dimIn * dimOut82]byte) {
+
+	offset := 0
+	for c := inputOffset; c < inputOffset+dimIn; c++ {
+		for iRow := outputOffset; iRow < outputOffset+dimOut82; iRow++ {
+			if c < len(matrixRows[iRow]) {
+				coeff := matrixRows[iRow][c]
+				copy(matrix[offset*32:], mulTableLow[coeff][:])
+				copy(matrix[offset*32+16:], mulTableHigh[coeff][:])
+			} else {
+				// coefficients not used for this input shard (so null out)
+				copy(matrix[offset*32:], zeroCoeff[:])
+				copy(matrix[offset*32+16:], zeroCoeff[:])
+			}
+			offset += dimIn
+			if offset >= dimIn*dimOut82 {
+				offset -= dimIn*dimOut82 - 1
+			}
+		}
+	}
+}
+
+// Construct block of matrix coefficients for 4 outputs rows in parallel
+func setupMatrix84(matrixRows [][]byte, inputOffset, outputOffset int, matrix *[(16 + 16) * dimIn * dimOut84]byte) {
+
+	offset := 0
+	for c := inputOffset; c < inputOffset+dimIn; c++ {
+		for iRow := outputOffset; iRow < outputOffset+dimOut84; iRow++ {
+			if c < len(matrixRows[iRow]) {
+				coeff := matrixRows[iRow][c]
+				copy(matrix[offset*32:], mulTableLow[coeff][:])
+				copy(matrix[offset*32+16:], mulTableHigh[coeff][:])
+			} else {
+				// coefficients not used for this input shard (so null out)
+				copy(matrix[offset*32:], zeroCoeff[:])
+				copy(matrix[offset*32+16:], zeroCoeff[:])
+			}
+			offset += dimIn
+			if offset >= dimIn*dimOut84 {
+				offset -= dimIn*dimOut84 - 1
+			}
+		}
+	}
+}
+
+// Invoke AVX512 routine for 2 output rows in parallel
+func galMulAVX512Parallel82(in, out [][]byte, matrixRows [][]byte, inputOffset, outputOffset int) {
+
+	done := len(in[0])
+	if done > 0 {
+		inputEnd := inputOffset + dimIn
+		if inputEnd > len(in) {
+			inputEnd = len(in)
+		}
+		outputEnd := outputOffset + dimOut82
+		if outputEnd > len(out) {
+			outputEnd = len(out)
+		}
+
+		matrix82 := [(16 + 16) * dimIn * dimOut82]byte{}
+		setupMatrix82(matrixRows, inputOffset, outputOffset, &matrix82)
+		addTo := inputOffset != 0 // Except for the first input column, add to previous results
+		_galMulAVX512Parallel82(in[inputOffset:inputEnd], out[outputOffset:outputEnd], &matrix82, addTo)
+
+		done = (done >> 6) << 6
+	}
+	remain := len(in[0]) - done
+
+	if remain > 0 {
```

Comment:
nit: let's just return, so we can kill some indentation.

Reply:
Done.